test/javaTests/src/org/criu/java/tests/CheckpointRestore.java

```diff
@@ -0,0 +1,345 @@
+	 * Get the pid of the test process
+	 */
+
+	pid = getPid(testName);
+	try {
+		/*
+		 * Checkpoint and Restore the process, the checkpoint function
+		 * calls the Restore function.
+		 */
+		checkpoint(pid);
+
+	} catch (Exception e) {
+		/*
+		 * If exception occurs put the MappedByteBuffer to Helper.STATE_TERMINATE-Terminate state.
+		 * On reading the terminate state, the test process terminates, else it
+		 * may go on looping.
+		 */
+		mappedByteBuffer.putChar(Helper.MAPPED_INDEX, Helper.STATE_TERMINATE);
+		Assert.fail(testName + ": Exception occurred while during checkpointing" + e, e);
+	}
+	/*
+	 * Wait for test process to finish
+	 */
+	currentState = mappedByteBuffer.getChar(Helper.MAPPED_INDEX);
+	while (Helper.STATE_RESTORE == currentState) {
+		currentState = mappedByteBuffer.getChar(Helper.MAPPED_INDEX);
+	}
+
+	/*
+	 * If a test passes it puts the MappedByteBuffer to Helper.STATE_PASS-Pass state,
+	 * On failing to Helper.STATE_FAIL-Fail state, and if our Buffer is in Helper.STATE_TERMINATE state
+	 * its because the checkpoint-restore of test process failed.
+	 */
+
+	Assert.assertNotEquals(currentState, Helper.STATE_TERMINATE, testName + ": ERROR: Checkpoint-Restore failed");
+	Assert.assertNotEquals(currentState, Helper.STATE_FAIL, testName + ": ERROR: Test Failed, Check Log for details");
+	Assert.assertEquals(currentState, Helper.STATE_PASS, testName + " ERROR: Unexpected State of Mapped Buffer");
+	System.out.println("-----" + "PASS" + "-----");
+
+	}
+
+	/**
+	 * Remove .img files, stats-dump and stats-restore files from Log Directory
+	 *
+	 * @throws IOException
+	 */
+	@AfterTest
+	void cleanup() throws IOException {
+	int i;
+	String currentPath = System.getProperty("user.dir");
+	currentPath = currentPath + "/" + logFolder;
+	File deleteFile;
+	File dir = new File(currentPath);
+	String[] logFiles = dir.list(new ImgFilter());
+	if (null != logFiles) {
+		for (i = 0; i < logFiles.length; i++) {
+		deleteFile = new File(currentPath + logFiles[i]);
+		Files.delete(deleteFile.toPath());
+		}
+	}
+
+	boolean exists = Files.exists(Paths.get(currentPath + "stats-dump"));
+	if (exists) {
+		Files.delete(Paths.get(currentPath + "stats-dump"));
+	}
+
+	exists = Files.exists(Paths.get(currentPath + "stats-restore"));
+	if (exists) {
+		Files.delete(Paths.get(currentPath + "stats-restore"));
+	}
+	}
+
+	/**
+	 * Checkpoints the process, if process has been checkpointed correctly calls restore,
+	 * else makes a copy of dump.log, puts MappedBuffer to 'terminate' state and marks
+	 * test as failed
+	 *
+	 * @param pid Pid of process to be checkpointed
+	 * @throws IOException
+	 * @throws InterruptedException
+	 */
+	private void checkpoint(String pid) throws IOException, InterruptedException {
+	System.out.println("Checkpointing process " + pid);
+
+	ProcessBuilder builder = new ProcessBuilder("criu", "dump", "--tcp-established", "--shell-job", "-t", pid, "-vvv", "-D", logFolder, "-o", "dump.log");
```

Comment:
It would be better to specify CRIU options per test case. Also we should be testing with the compiled `criu` executable not the one installed on the system.

Reply:
CRIU options are now specified per test case, and the compiled CRIU is being used.